_posts/2021-01-09-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,629 @@
+But thanks to this work, JuliaInterpreter, and thus Revise, now load in a flash.
+
+A full discussion on the invalidations work can be found in [this blog post](https://julialang.org/blog/2020/08/invalidations/).
+
+### Per-Module Optimization Flags
+>The compiler optimization level can now be set per-module using the experimental macro Base.Experimental.@optlevel n. For code that is not performance-critical, setting this to 0 or 1 can provide significant latency improvements ([#34896](https://github.com/JuliaLang/julia/issues/34896)).
+
+> Compilation and type inference can now be enabled or disabled at the module level using the experimental macro Base.Experimental.@compiler_options ([#37041](https://github.com/JuliaLang/julia/issues/37041)).
+
+Recall that I said most of Julia compilation  isn't even that useful for plotting, since making it run fast isn't a priority (but loading fast is).
+
+`@optlevel` only controls which LLVM optimization passes run; which is right at the end of the compilation pipeline.
+Turning off compilation and type inference on the other hand turns off a ton more.
+
+
+## Internals
+### Manually Created Back-edges for Lowered Code Generated Functions
+This is a very niche and not really at all user-facing feature.
+To understand why this matters, it's worth understanding how Cassette works.
+I wrote a [blog post on this a few years ago](https://invenia.github.io/blog/2019/10/30/julialang-features-part-1#making-cassette).
+
+In quite the opposite, Julia 1.3 allowed for the creation of more invalidations through allowing back-edges to be manually attached to the `CodeInfo` for `@generated` functions that return lowered code.
+Back-edges are the connections from methods back to each method instance that calls them.
+This is what allows invalidations to work, as when a method is redefined, it needs to know what things to recompile.
+This change allowed those back-edges to be manually specified for `@generated` functions that were working at the lowered code level.
+This is useful since this technique is primarily used for generating code based on the (lowered) code of existing methods.
+For example, in [Zygote](https://github.com/FluxML/Zygote.jl), generating the gradient code from the code of the primal method.
+So you want to be able to trigger the regeneration of this code when that original method changes.
+
+
+Basically, this allows code that uses [Cassette.jl](https://github.com/JuliaLabs/Cassette.jl), and [IRTools.jl](https://github.com/MikeInnes/IRTools.jl) to not suffer from [#265](https://github.com/JuliaLang/julia/issues/265)-like problems.
+A particular case of this is for [Zygote](https://github.com/FluxML/Zygote.jl), where redefining a function called by the code that was being differentiated did not result in an updated gradient (unless `Zygote.refresh()`) was run.
+This was annoying for
```

Comment:
Incomplete sentence?

Reply:
```suggestion
This was annoying for working in the REPL, where you might e.g. change your neural nets loss function without redefining the network.
In that case, if you forgot to run `Zyote.refresh()` it would just train wrong.
```